### ISSUE TITLE
Equation alignment isn't tabular 

### ISSUE THREAD

**Issue by SUPERCILEX** [typst/typst#695]
```
$
"here" & "there" \
"something really long"
$
```

![image](https://user-images.githubusercontent.com/9490724/230797254-96dac2a0-8c2f-4409-910b-0fcbf87d2357.png)

An `&` be added implicitly to the end of `"something really long"` so it renders like this:

![image](https://user-images.githubusercontent.com/9490724/230797263-5a1e2977-3f6f-40ea-8e20-3c68edbac38c.png)

**Comment by laurmaedje** [comment]
What do you mean by "implicit break"?

**Comment by laurmaedje** [comment]
I'm open to changing this, but why is it better that way? If you want to align at that point, why not put the `&` there?

**Comment by SUPERCILEX** [comment]
Because in a long equation with a dozen steps, it'd be pretty annoying to have to put the & everywhere when you only need it once. I'd also argue that the current behavior doesn't make sense because adding a single & is equivalent to not adding it at all since it doesn't do anything until you add another one. And then once you add another &, the two & will only align with each other instead of the whole equation which kind of makes them pointless IMO. So it seems like you either want no & or every line to have an & which is what the Latex behavior provides.

**Comment by owiecc** [comment]
Wouldn't an implicit alignment mark be better on the left hand side (LHS)? 
```
$
y &= (a+b)^2 \
= (a+b)(a+b) \
= aa + ab + ab + bb \
= a^2 + 2ab + b^2
$
```

**Comment by SUPERCILEX** [comment]
I don't see how that could work without treating = and other characters specially. I've updated the issue title to better reflect what this is about: the idea is to treat & as column delimiters in an implicit table. I think it's reasonable to have this be a parameter of the equation function as people will probably want to stick with one style.

**Comment by owiecc** [comment]
You do not need to treat any characters specially. If there exist a line with `&` then the assumption is that lines should be somehow aligned. If no alignment symbol is present in a line then add the symbol before the first character in the line. 
```
$
y = &(a+b)^2 \
(a+b)(a+b) \
aa + ab + ab + bb \
a^2 + 2ab + b^2
$
```

**Comment by SUPERCILEX** [comment]
Hmmm, I think this is still a no from me because if you add alignment to the line later (see example below), you have to go back and add alignment to the beginning which is inconsistent and confusing.

```
... &= ... \
= ... & "By theorem foo"
```

**Comment by owiecc** [comment]
> the idea is to treat & as column delimiters in an implicit table

Now I understand better what you'd like to accomplish and I have some comments. 

For a *tabular math* layout, like https://tex.stackexchange.com/a/246477 it would probably be better to set with `#table` with an ability to align the individual equations in table columns with `&` symbol. It looks like a more idiomatic way to describe the layout. If the math environment allows a tabular alignment I am afraid that we will see a lot of early web like behaviour of laying out everything in a table. I don't think it is a good idea. We did this because this was our only option. 

In your *specific* example I think a non-tabular layout would be better. You are visually separating two related contents (equation line and some comment) by an *unknown* (possibly very wide) amount of space. This is what tables of content typically look like. Titles on the left and pages on the right. Wouldn't Tufte's approach be better in this case? See [TOC layout example](https://tex.stackexchange.com/a/631408), just imagine titles as equations and page numbers as comments. A table of content is then a list of content. Much more readable. Maybe wrap the in a function so you can have a common styling. 

```
... &= ... \
&= ... eqncomment("By theorem foo")
```

**Comment by SUPERCILEX** [comment]
> with an ability to align the individual equations in table columns with & symbol.

If I'm reading this right, then this is what I'm proposing. This is also how `\being{align*}` works in latex.

> I am afraid that we will see a lot of early web like behaviour of laying out everything in a table. I don't think it is a good idea. We did this because this was our only option.

I don't understand this.

> Wouldn't Tufte's approach be better

No? The bottom screenshot looks way better IMO because there is a clear separation of content: left is titles, right is page numbers. The first screenshot is a visual mess and is how typst currently works unless you put & at the end of every line. I'm basically trying to get the bottom screenshot by default.

**Comment by owiecc** [comment]
> If I'm reading this right, then this is what I'm proposing.

Yes, I just argue about the implementation. I believe some form of tabular layout is necessary. However, adding an option to align equations in a table would be much better IMHO then adding a table alignment mode to equation environment. 

> I don't understand this.

I refer to the time before CSS. Any web page layout was done using tables. Headers, menus, … you name it. Any kind of page layout had to be massaged into a table. So things that were not tables were made to be tables. Everything was a table. 

> The bottom screenshot looks way better IMO because there is a clear separation of content

You are disagreeing with some big people here. See Bringhurst [The Elements of Typographic Style](https://en.wikipedia.org/wiki/The_Elements_of_Typographic_Style) (known as Typographers' Bible) chapter 2.1.10 on horizontal spacing. Some parts are quoted in [The Elements of Typographic Style Applied to the Web](http://webtypography.net/2.1.10) and Tufte [The Visual Display of Quantitative Information](https://www.edwardtufte.com/tufte/books_vdqi) (data visualization guru).

In short, some spacing, colour or font variant is enough to distinguish different content type but too much spacing brakes the composition.

**Comment by SUPERCILEX** [comment]
> I believe some form of tabular layout is necessary. However, adding an option to align equations in a table would be much better IMHO then adding a table alignment mode to equation environment.

Gotya, I guess I don't understand the difference. Would you mind clarifying further?

> I refer to the time before CSS. [...] Everything was a table.

Haha, ok. I didn't even exist when CSS was invented. :) Anyway, for equations I think aligning them in columns (i.e. as a table) is perfectly reasonable.

**Comment by SUPERCILEX** [comment]
> I believe some form of tabular layout is necessary. However, adding an option to align equations in a table would be much better IMHO then adding a table alignment mode to equation environment.

Gotya, I guess I don't understand the difference. Would you mind clarifying further?

> I refer to the time before CSS. [...] Everything was a table.

Haha, ok. I didn't even exist when CSS was invented. :) Anyway, for equations I think aligning them in columns (i.e. as a table) is perfectly reasonable. I think this is where we disagree.

> You are disagreeing with some big people here.

Huh, wasn't aware. Just to be clear, I'm not advocating that contents be right aligned (the content should be as far left as it can be without breaking a column). I agree that too much whitespace means you can't follow which line a piece of content applies to, but I simply don't see how the first is better than the second:
![image](https://user-images.githubusercontent.com/9490724/232164998-9e82306e-49ac-47a6-8b0c-3acfc5729743.png)

**Comment by Enivex** [comment]
> If the math environment allows a tabular alignment I am afraid that we will see a lot of early web like behaviour of laying out everything in a table. I don't think it is a good idea. We did this because this was our only option. 

Tabular alignment in math is the defacto standard. I'm not sure what you believe the alternative is. Currently, math alignment is completely broken in Typst when you have more than two columns.

**Comment by owiecc** [comment]
An alternative:
```
#grid(
  columns: 2,
  $ a&=b $, $ c&=d $, 
  $ e&=f $, $ g&=h $, 
  $ k&=l $, $ m&=h n, 
)
```
Want a three column layout instead? Change the `columns` parameter to 3.

**Comment by Enivex** [comment]
And what exactly does that accomplish other than making math tedious to enter?

**Comment by owiecc** [comment]
Easier to control spacing. No fake spaces to add space, you can use gutters for that.

Easier to change layout. It is changed with parameters. Number of columns. Skipped elements. Try changing number of columns in a big latex `alignat`. Not fun. 

Easier to maintain the code. Math stays in math environment. Layout stays in layout code. 

Easier to reason about the code (for IDEs). No need to parse a quite complex string that possibly has missing line breaks, wrong number of cells, missing or duplicated alignment marks, etc. to figure out what is what. 

Easier to script. If you e.g. want all your math cells to align on a first plus sign you create a function that is passed to all grid elements prior to type setting that injects a `&` before the first plus.

**Comment by Enivex** [comment]
> Easier to control spacing. No fake spaces to add space, you can use gutters for that.

I fail to see why they're mutually exclusive. There's no reason why you cannot set column spacing in an equation.

> Easier to change layout. It is changed with parameters. Number of columns. Skipped elements. Try changing number of columns in a big latex `alignat`. Not fun.

You seem to have a very narrow view of how align is used. Most of the time it's not just a big grid of equations. I can think of very few situations where it would be desirable to "reorder" the entries.

> Easier to reason about the code (for IDEs). No need to parse a quite complex string that possibly has missing line breaks, wrong number of cells, missing or duplicated alignment marks, etc. to figure out what is what.

Usability for humans should never be sacrificed because we want to be nice to the parser.

Obviously I can see situations where it would be beneficial to use a grid of equations like you propose, but that should definitely not be the default way to align equations.
